dart/collision/CollisionDetector.h

```diff
@@ -43,169 +43,179 @@
   ///
-  /// The contact force acting on bodyNode2 is -force, which is the opposite
-  /// direction of the force.
-  Eigen::Vector3d force;
+  /// Internally, this function creates shared_ptr from unique_ptr returned from
+  /// createCollisionGroup() so the performance would be slighly worse than
+  /// using std::make_unique.
+  std::shared_ptr<CollisionGroup> createCollisionGroupAsSharedPtr();
 
-  /// First colliding body node
-  dynamics::WeakBodyNodePtr bodyNode1;
+  /// Create a collision group from any objects that are supported by
+  /// CollisionGroup::addShapeFramesOf().
+  ///
+  /// The objects can be any of ShapeFrame, std::vector<ShapeFrame>,
+  /// CollisionGroup, BodyNode, and Skeleton.
+  ///
+  /// Note that this function adds only the ShapeFrames of each object at the
+  /// moment of this function is called. The aftwerward changes of the objects
+  /// does not affect on this CollisionGroup.
+  template <typename... Args>
+  std::unique_ptr<CollisionGroup> createCollisionGroup(const Args&... args);
 
-  /// Second colliding body node
-  dynamics::WeakBodyNodePtr bodyNode2;
+  /// Helper function that creates and returns CollisionGroup as shared_ptr.
+  template <typename... Args>
+  std::shared_ptr<CollisionGroup> createCollisionGroupAsSharedPtr(
+      const Args&... args);
 
-  /// First colliding shape of the first body node
-  dynamics::ShapePtr shape1;
+  /// Perform collision detection for group.
+  virtual bool collide(CollisionGroup* group,
+                       const CollisionOption& option, CollisionResult& result) = 0;
 
-  /// Second colliding shape of the first body node
-  dynamics::ShapePtr shape2;
+  /// Perform collision detection for group1-group2.
+  virtual bool collide(CollisionGroup* group1, CollisionGroup* group2,
+                       const CollisionOption& option, CollisionResult& result) = 0;
 
-  /// Penetration depth
-  double penetrationDepth;
+protected:
 
-  // TODO(JS): triID1 will be deprecated when we don't use fcl_mesh
-  /// \brief
-  int triID1;
+  class CollisionObjectManager;
+  class NoneSharingCollisionObjectManager;
+  class SharingCollisionObjectManager;
 
-  // TODO(JS): triID2 will be deprecated when we don't use fcl_mesh
-  /// \brief
-  int triID2;
+  /// Constructor
+  CollisionDetector() = default;
 
-  // TODO(JS): userData is an experimental variable.
-  /// \brief User data.
-  void* userData;
-};
+  /// Claim CollisionObject associated with shapeFrame. New CollisionObject
+  /// will be created if it hasn't created yet for shapeFrame.
+  std::shared_ptr<CollisionObject> claimCollisionObject(
+      const dynamics::ShapeFrame* shapeFrame);
 
-/// \brief class CollisionDetector
-class CollisionDetector
-{
-public:
-  /// \brief Constructor
-  CollisionDetector();
+  /// Create CollisionObject
+  virtual std::unique_ptr<CollisionObject> createCollisionObject(
+      const dynamics::ShapeFrame* shapeFrame) = 0;
```

Comment:
Yes. The main motivation behind this is to make `CollisionDetector` create single `CollisionObject` per `ShapeFrame` using RAII.

`claimCollisionObject(~)` returns `std::shared_ptr<CollisionObject>`. The returning pointer is either of one created from `createCollisionObject` if it's not created for the passed in `ShapeFrame` or one already created for the `ShapeFrame`. The management of the uniquess of `CollisionObject` per `ShapeFrame` is done using `std::map<ShapeFrame*, std::weak_ptr<CollisionObject>` in `CollisionDetector` (more precisely, in `CollisionDetector::SharingCollisionObjectManager`). The `CollisionObject` should be removed from the map when it's not shared by any `CollisionObject`. However, the thing is `std::shared_ptr<CollisionObject>` will be automatically destructed when it's not referenced by any `CollisionGroup` (as the map holds `CollisionObject` as weak pointer), but `CollisionDetector` doesn't know when it's happening.

To handle this, a custom deleter of shared pointer is used where it removes `CollisionObject` from the map before it deletes `CollisionObject`. I could make `createCollisionObject` create `shared_ptr` with the custom deleter by passing it from `CollisionDetector` to each concrete collision detector, but also wondered if it's a good practice. To me, it seemed the management of collision object management should be totally the responsibility of `CollisionDetector` and the each collision detector just should create the concrete collision object (e.g., `FCLCollisionObject`).

I'm open to hearing better solution here.